test/work_submitter.test.ts

```diff
@@ -18,141 +19,256 @@ jest.mock('../src/utils/logger.js', () => ({
+    mockedSorobanHelper.loadAuction.mockResolvedValue(undefined);
+    mockedSorobanHelper.submitTransaction.mockRejectedValue(
+      new ContractError(ContractErrorType.InvalidLiqTooLarge)
+    );
+
+    const submission = {
+      type: WorkSubmissionType.LiquidateUser,
+      user: Keypair.random().publicKey(),
+      liquidationPercent: BigInt(1),
+    };
+
+    const result = await workSubmitter.submit(submission);
+    expect(result).toBe(false);
+    expect(submission.liquidationPercent).toBe(BigInt(1));
+    expect(logger.error).toHaveBeenCalled();
+    expect(mockedSendSlackNotif).toHaveBeenCalled();
+  });
+
+  it('does not adjust fill percentage for general error', async () => {
+    mockedSorobanHelper.loadAuction.mockResolvedValue(undefined);
+    mockedSorobanHelper.submitTransaction.mockRejectedValue(
+      new ContractError(ContractErrorType.InvalidLiquidation)
+    );
+
+    const submission = {
+      type: WorkSubmissionType.LiquidateUser,
+      user: Keypair.random().publicKey(),
+      liquidationPercent: BigInt(50),
+    };
+
+    const result = await workSubmitter.submit(submission);
+    expect(result).toBe(false);
+    expect(submission.liquidationPercent).toBe(BigInt(50));
+    expect(logger.error).toHaveBeenCalled();
+    expect(mockedSendSlackNotif).toHaveBeenCalled();
+  });
+
+  it('should adjust fill percentage based on contract error on each retry', async () => {
+    mockedSorobanHelper.loadAuction.mockResolvedValue(undefined);
+    mockedSorobanHelper.submitTransaction.mockRejectedValue(
+      new ContractError(ContractErrorType.InvalidLiqTooSmall)
+    );
+
+    const submission = {
+      type: WorkSubmissionType.LiquidateUser,
+      user: Keypair.random().publicKey(),
+      liquidationPercent: BigInt(50),
+    };
+
+    workSubmitter.addSubmission(submission, 3);
+    while (workSubmitter.processing) {
+      await new Promise((resolve) => setTimeout(resolve, 50));
+    }
+    // 3 retries plus the final increment before dropping
+    expect(submission.liquidationPercent).toBe(BigInt(54));
+    expect(logger.error).toHaveBeenCalledWith(
+      expect.stringContaining('Dropped liquidation for user')
+    );
+  });
+
   it('should submit a bad debt transfer successfully', async () => {
     const submission: WorkSubmission = {
       type: WorkSubmissionType.BadDebtTransfer,
-      user: 'testUser',
+      user: Keypair.random().publicKey(),
     };
 
     const result = await workSubmitter.submit(submission);
-    const expectedLogMessage = `Successfully submitted bad debt transfer for user: ${submission.user}`;
     expect(result).toBe(true);
     expect(mockedSorobanHelper.submitTransaction).toHaveBeenCalled();
-    expect(logger.info).toHaveBeenCalledWith(expectedLogMessage);
-    expect(sendSlackNotification).toHaveBeenCalledWith(expectedLogMessage);
+    expect(logger.info).toHaveBeenCalled();
+    expect(sendSlackNotification).toHaveBeenCalled();
```

Comment:
omitted validation of log messages in favor of just validating that logs are created in general.